Plugin Doc and Example Refactors for #5596 (#5621)
ethereum/web3.js: docs/docs/guides/web3_plugin_guide/plugin_authors.md

This guide intends to provide the necessary context for developing plugins for web3.js.

:::caution

> To provide type safety and hinting for your plugin users, please refer to the [Setting Up Module Augmentation](/docs/guides/web3_plugin_guide/plugin_authors#setting-up-module-augmentation) section for how to augment the `Web3Context` module to enable typing features for your plugin.

Reviewer: Tiny suggestion:
```suggestion
To provide type safety and IntelliSense for your plugin users, please refer to the [Setting Up Module Augmentation](/docs/guides/web3_plugin_guide/plugin_authors#setting-up-module-augmentation) section for how to augment the `Web3Context` module to enable typing features for your plugin.
```

```typescript
// registering_a_plugin.ts
import { Web3Context } from 'web3-core';

import { CustomRpcMethodsPlugin } from './custom_rpc_methods_plugin';

const web3Context = new Web3Context();

> web3Context.registerPlugin(new CustomRpcMethodsPlugin());

Reviewer: This code snippet and its section seem to need to be moved up before the title `Using the Inherited Web3Context`, doesn't it?

Author: No it's meant to be where it is. It's adding additional context for the plugin author showing how their code could be affected by what the user does/n't do (e.g. this example shows a provider error being thrown from the plugin author's code because the plugin user didn't provide a provider). [This](https://github.com/web3/web3.js/pull/5621/commits/96c90445070f2751a1286597cd40fdc24b3d421e) commit does reword this section a little

	this._contract.link(parentContext);
}
```

## Setting Up Module Augmentation

> In order to provide typing support for your plugin when it's registered by the user, we must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, we're making TypeScript aware that we are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of your plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).

Reviewer: I think there is a mixed "we" and "you" here.
So either using "you" and "your":
```suggestion
In order to provide type safety and IntelliSense for your plugin when it's registered by the user, you must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, you will be making TypeScript aware that you are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of your plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).
```

Or using "we" and "our" if we refer to the plugin user as we are talking about ourselves (but it is not a good suggestion I think because it would confuse the reader):
```suggestion
In order to provide type safety and IntelliSense for our plugin when it's registered by the user, we must [augment](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) the `Web3Context` module. In simpler terms, we're making TypeScript aware that we are modifying the interface of `Web3Context`, and any class that extends it, to include the interface of our plugin (i.e. your plugin's added methods, properties, etc.). A good tutorial that further explains the topic can be found [here](https://www.digitalocean.com/community/tutorials/typescript-module-augmentation).
```

### A Quick Disclaimer

> The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add on your plugin's additional functionality to it's interface. By augmenting `Web3Context` to include your plugin's interface, we're essentially providing a blanket augmentation that adds you plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).

Reviewer: Here also there are "... your plugin's ..." and "...we're...".

```suggestion
The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add your plugin's additional functionality to its interface. And, by augmenting `Web3Context` to include your plugin's interface, you're essentially providing a blanket augmentation that adds your plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).
```

And actually besides using "you" is more clear, if "we" is used it may confuse the reader that this would be done, or already done, by web3.js team.

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/92402f68f7ab0d0102a495714fd6188ac5e04c2e) commit

	public someMethod() {
		return 'someValue';
	}
}

// Module Augmentation
declare module './reexported_web3_context' {
	interface Web3Context {
		customRpcMethods: CustomRpcMethodsPlugin;
	}
}

export { Web3Context };
```

The following represent what your plugin users would see:

![web3 context augmentation side effect](./assets/web3_context_augmentation_side_effect.png 'web3Context augmentation side effect')

The above screenshot shows intellisense thinking `.customRpcMethods.someMethod` is available to call on the instance of `Web3`, even though we haven't registered `CustomRpcMethodsPlugin` - running this code would result in an error.
:::

### Re-exporting Web3Context

Currently TypeScript's module augmentation only supports named exports, so the first step in augmenting `Web3Context` is to re-export it as a named export. To do this we're going to create a `reexported_web3_context.ts` file (the name of this file can be whatever, but for the sake of this guide, we're going to assume it's named `reexported_web3_context.ts` and is located within the same directory as our `custom_rpc_methods_plugin.ts` file). The file contents should be as follows:

> ```typescript

Reviewer: ```suggestion
```typescript
// reexported_web3_context.ts
```

import { Web3Context } from 'web3-core';

export { Web3Context };
```

### Re-declaring the Module

> Now we're going to tell TypeScript that we're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, we're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.

Reviewer: ```suggestion
Now you're going to tell TypeScript that you're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, you're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.
```

> We start with the following:

Reviewer: I think it is not the "start" here. I think it is the main logic or the main file that contains the plugin class.

```suggestion
Here is the content of the source file that contains the main plugin class:
```
(This would also remove "we" to keep with the context)

Reviewer 2: Addressed bu [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

```typescript
// custom_rpc_methods_plugin.ts
import { Web3PluginBase } from 'web3-core';

> // Here we are importing our re-exported Web3Context
> // from the above section

Reviewer: ```suggestion
// Here is the import of the re-exported `Web3Context` 
// that has been mentioned in the above section.
```
(This would also remove "we" to keep with the context)

Reviewer 2: Addressed bu [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

import { Web3Context } from './reexported_web3_context';

export class CustomRpcMethodsPlugin extends Web3PluginBase {
	public pluginNamespace = 'customRpcMethods';

	public someMethod() {
		return 'someValue';
	}
}

> // Here is where we are beginning to augment the Web3Context module

Reviewer: ```suggestion
// Here is the augmentation of the Web3Context module
```
(This would also remove "we" to keep with the context)

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/4cb635cb4a8489aaf3d0cd3b6479aad3764ed20e) commit

declare module './reexported_web3_context' {...}
```

### Adding our Plugin's Interface

> Now that TypeScript's aware that the interface of the `reexport_web3_context` module is going to be augmented, we add our changes. In this case, we're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what we're going to be calling `.registerPlugin` on:

Reviewer: Actually, "TypeScript`s" is not the same as "TypeScript is".  We can not abbreviate "is" with names. With names, the " 's " is used to show possession and ownership.

Also using you to keep with the context.

```suggestion
Now that TypeScript is aware that the interface of the `reexport_web3_context` module is going to be augmented, you can add our changes. In this case, you're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what the **plugin-user** is going to be calling `.registerPlugin` on:
```

	public pluginNamespace = 'customRpcMethods';

	public someMethod() {
		return 'someValue';
	}
}

declare module './reexported_web3_context.ts' {
	// Here is where we're adding our plugin's
	// interface to the interface of Web3Context
	interface Web3Context {
		customRpcMethods: CustomRpcMethodsPlugin;
	}
}
```

:::info
The property name (i.e. `pluginNamespace`), `customRpcMethods` in

```typescript
{
	customRpcMethods: CustomRpcMethodsPlugin;
}
```

**MUST** be the same as the `pluginNamespace` set by the plugin.

```typescript
import { Web3PluginBase } from 'web3-core';

export class CustomRpcMethodsPlugin extends Web3PluginBase {
	public pluginNamespace = 'customRpcMethods';

	...
}
```

This is because `.registerPlugin` will use the `pluginNamespace` property provided by the plugin as the property name when it registers the plugin with the class instance you call `.registerPlugin` on:

```typescript
const web3 = new Web3('http://127.0.0.1:8545');
web3.registerPlugin(new CustomRpcMethodsPlugin());
// Now customRpcMethods (i.e. the pluginNamespace) is available
// on our instance of Web3
web3.customRpcMethods;
```

:::

### Exporting Our Augmented Web3Context

> Lastly we just need to export our augmented `Web3Context` by adding the following after our module re-declaration:

Reviewer: To keep with the context.
```suggestion
Lastly, you just need to export our augmented `Web3Context` by adding the following after our module re-declaration:
```

```typescript
export { Web3Context };
```

The full code example is as follows:

```typescript
// custom_rpc_methods_plugin.ts
import { Web3PluginBase } from 'web3-core';

import { Web3Context } from './custom_rpc_methods_plugin';

export class CustomRpcMethodsPlugin extends Web3PluginBase {
	public pluginNamespace = 'customRpcMethods';

	public someMethod() {
		return 'someValue';
	}
}

declare module './custom_rpc_methods_plugin.ts' {
	interface Web3Context {
		simplePlugin: SimplePlugin;
	}
}

> // Here is where we are exporting our augmented Web3Context

Reviewer: To keep with the context:
```suggestion
// Here is where you are exporting your augmented Web3Context
```

### A Quick Disclaimer

The `registerPlugin` method exists on the `Web3Context` class, so any class that `extends Web3Context` has the ability to add your plugin's additional functionality to its interface. By augmenting `Web3Context` to include your plugin's interface, you're essentially providing a blanket augmentation that adds your plugin's interface to **all** Web3 modules that extend `Web3Context` (i.e. `web3`, `web3-eth`, `web3-eth-contract`, etc.).

:::warning
By augmenting `Web3Context` (and by extension all class interfaces that extend it), your plugin's interface will show up in things like intellisense for **all** Web3 modules that extend `Web3Context`, even if your plugin isn't registered - This is something worth making your users aware of, as they'll only be able to use your plugin if they register it with a Web3 class instance using `.registerPlugin`

> For context, here is a code example of this augmentation side-effect (the code in this example is further explained in the subsequent sections):

Reviewer: I would like to know what the others think about this also :smile:. Especially our other approver: @avkos 

```suggestion
For context, here is a code example of this augmentation that happens regardless of whether `registerPlugin` is called or not (the code in this example is further explained in the subsequent sections):
```

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/921ab8816ea0578e8b51abd23e1f2a22df24fbef) commit

```typescript
// custom_rpc_methods_plugin.ts
import { Web3PluginBase } from 'web3-core';

import { Web3Context } from './reexported_web3_context';

export class CustomRpcMethodsPlugin extends Web3PluginBase {
	public pluginNamespace = 'customRpcMethods';

	public someMethod() {
		return 'someValue';
	}
}

// Module Augmentation
declare module './reexported_web3_context' {
	interface Web3Context {
		customRpcMethods: CustomRpcMethodsPlugin;
	}
}

export { Web3Context };
```

The following represent what your plugin users would see:

> ![web3 context augmentation side effect](./assets/web3_context_augmentation_side_effect.png 'web3Context augmentation side effect')

Reviewer: Here also, I would like to know what the others think about this also :smile:. Especially our other approver: @avkos 
```suggestion
![web3 context augmentation](./assets/web3_context_augmentation.png 'web3Context augmentation')
```

(The file name would need to be also modified accordingly.)

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/921ab8816ea0578e8b51abd23e1f2a22df24fbef) commit

> ![web3 context augmentation side effect](./assets/web3_context_augmentation.png 'web3Context augmentation side effect')

Reviewer: Here is a tiny missed modification, I think:
```suggestion
![web3 context augmentation](./assets/web3_context_augmentation.png 'web3Context augmentation')
```

import { Web3Context } from 'web3-core';

export { Web3Context };
```

### Re-declaring the Module

Now you're going to tell TypeScript that you're interested in re-defining a module's (in this case `reexported_web3_context`) interface. In simpler terms, TypeScript is already aware of what methods and classes exist for each web3.js module, but when registering a plugin, you're adding additional methods and/or classes to the module's interface and TypeScript needs a little help understanding what's going to be available within the module after the plugin is registered.

```typescript
// custom_rpc_methods_plugin.ts
import { Web3PluginBase } from 'web3-core';

// Here the re-exported Web3Context from
// the previous section is being imported
import { Web3Context } from './reexported_web3_context';

export class CustomRpcMethodsPlugin extends Web3PluginBase {
	public pluginNamespace = 'customRpcMethods';

	public someMethod() {
		return 'someValue';
	}
}

// Here is the declaration to TypeScript that you are
// augmenting the imported module (i.e. ./reexported_web3_context)
declare module './reexported_web3_context' {...}
```

### Adding Your Plugin's Interface

> Now that TypeScript is aware that the interface of the `reexport_web3_context` module is going to be augmented, you can add your plugin's interface. In this case, you're adding the interface of `SimplePlugin` to the interface of `Web3Context` which is what the **plugin-user** is going to be calling `.registerPlugin` on:

Reviewer: The name `SimplePlugin` here is not used elsewhere in this section. However, the one used in the same section below is  `CustomRpcMethodsPlugin`. 
Actually, there is a mixed usage of the name of the plugin in other sections as well. And I think this would confuse the reader. So could you unify the name across all sections and code snippets, except if there is a need for a specific reason?

Author: There's no need, it's a copy paste error and they should all be CustomRpcMethodsPlugin

Author: Addressed by [this](https://github.com/web3/web3.js/pull/5621/commits/63ee5f86c0a47d3eeac40c4dc101caf4a1540b46) commit
